lib/ansible/plugins/filter/core.py

```diff
@@ -292,14 +292,14 @@ def get_encrypted_password(password, hashtype='sha512', salt=None):
             if sys.platform.startswith('darwin'):
                 raise AnsibleFilterError('|password_hash requires the passlib python module to generate password hashes on Mac OS X/Darwin')
             saltstring = "$%s$%s" % (cryptmethod[hashtype], salt)
-            encrypted = crypt.crypt(password, saltstring)
+            encrypted = crypt.crypt(str(password), saltstring)
         else:
             if hashtype == 'blowfish':
                 cls = passlib.hash.bcrypt
             else:
                 cls = getattr(passlib.hash, '%s_crypt' % hashtype)
 
-            encrypted = cls.encrypt(password, salt=salt)
+            encrypted = cls.encrypt(str(password), salt=str(salt))
```

Comment:
This should also be `to_text` from ` ansible.module_utils._text` instead of `str` for python3 compatibility.

Reply:
Experimenting here shows that salt must be ascii-only text-type on python3. password can be either text or bytes on both python2 or python3.  I'd suggest:
``` python
# passlib allows either text or byte strings for password but salt must be ascii-only text on python3
encrypted = cls.encrypt(to_text(password, errors='surrogate_or_strict'), salt=to_text(salt, errors='surrogate_or_strict'))
```